### PR TITLE
store spotify tokens in mongo db

### PR DESCRIPTION
**Pull request by KimLinhVu**
What
===
* switch from storing spotify access and refresh tokens in local storage to db when user logs in to spotify account
* retrieve access token from db when making requests to spotify api
* update access token in db when token expires 

Why
===
* more safe than storing in local storage 

Who
===
@makushline @qibolun 

Test Plan
===
Spotify Tokens Schema: 
Screen Shot 2022-08-02 at 9 21 21 AM

### CODE REVIEW

File: server/models/SpotifyToken.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+const mongoose = require('mongoose')
+
+const SpotifyTokenSchema = new mongoose.Schema({
+  userId: String,
+  spotifyId: String,
+  accessToken: String,
+  refreshToken: String,
+  expiresIn: Number,
+  timeStamp: String
```
Comment (makushline) [inline server/models/SpotifyToken.js]: Shouldn't this be a date?

File: server/routes/spotify.js [hunk 2]
```diff
@@ -48,6 +54,31 @@ router.get('/login', (req, res) => {
...
+
+  if (!username) {
+    return next(new BadRequestError('Missing username field.'))
+  }
+  if (!password) {
+    return next(new BadRequestError('Missing password field.'))
+  }
+
+  /* decrypt hashed password */
+  /* search db for pw that matches correct user */
```
Comment (makushline) [inline server/routes/spotify.js]: I don't think this comment is correct. You're just validating that the password matches for the given user not searching the entire db for a pw that matches. If that were the case that would be a security vulnerability.

File: server/routes/spotify.js [hunk 3]
```diff
@@ -72,12 +103,25 @@ router.get('/callback', (req, res, next) => {
...
+
+            Spotify.locals.spotifyId = spotifyId
+            /* store tokens in database */
+            const found = await SpotifyToken.findOne({ userId, spotifyId })
+            if (found) {
+              await SpotifyToken.findOneAndUpdate({ userId, spotifyId }, {
+                accessToken: access_token, refreshToken: refresh_token, expiresIn: expires_in, timeStamp: Date.now()
+              })
+            } else {
+              const newTokens = new SpotifyToken({ userId, spotifyId, accessToken: access_token, refreshToken: refresh_token, expiresIn: expires_in, timeStamp: Date.now() })
```
Comment (makushline) [inline server/routes/spotify.js]: Nit: newToken singular.

File: server/routes/spotify.js [hunk 4]
```diff
@@ -119,4 +163,38 @@ router.get('/refresh_token', (req, res) => {
     })
 })
 
+router.get('/tokens', jwt.verifyJWT, async (req, res, next) => {
+  try {
+    const userId = req.userId
+    const spotifyId = Spotify.locals.spotifyId
+    const tokens = await SpotifyToken.findOne({ userId, spotifyId })
```
Comment (makushline) [inline server/routes/spotify.js:169]: Nit token singular since you're calling findOne.

File: server/routes/users.js [hunk 5]
```diff
@@ -145,4 +145,59 @@ router.post('/remove-following', jwt.verifyJWT, async (req, res, next) => {
   }
 })
 
-module.exports = router
\ No newline at end of file
+router.post('/add-follower-favorite', jwt.verifyJWT, async (req, res, next) => {
```
Comment (makushline) [inline server/routes/users.js]: Where is this getting called? I also don't think this should be part of this PR.

**Review by makushline — CHANGES_REQUESTED** [review]

File: server/routes/spotify.js (hunk 2 above)
Comment (KimLinhVu) [reply server/routes/spotify.js]: Woops, idk why I commented this.

File: server/routes/spotify.js (hunk 4 above)
Comment (KimLinhVu) [reply server/routes/spotify.js:169]: I think I made it plural b/c the object returned has an access and refresh token.

File: server/routes/users.js (hunk 5 above)
Comment (KimLinhVu) [reply server/routes/users.js]: Yea, idk how this got in here.

File: server/models/SpotifyToken.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+const mongoose = require('mongoose')
+
+const SpotifyTokenSchema = new mongoose.Schema({
+  userId: String,
+  spotifyId: String,
+  accessToken: String,
+  refreshToken: String,
+  expiresIn: Number,
+  timeStamp: Number
```
Comment (makushline) [inline server/models/SpotifyToken.js]: This should be a date. Mongo has a date type and mongoose gives you access to it:
https://mongoosejs.com/docs/schematypes.html
https://www.tutorialspoint.com/mongodb/mongodb_datatype.htm

File: server/models/SpotifyToken.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+const mongoose = require('mongoose')
+
+const SpotifyTokenSchema = new mongoose.Schema({
+  userId: String,
+  spotifyId: String,
+  accessToken: String,
+  refreshToken: String,
+  expiresIn: Number,
```
Comment (makushline) [inline server/models/SpotifyToken.js:8]: A more standard way to represent this is to call it expiresAt and have it also be a date type.

File: server/routes/spotify.js (hunk 4 above)
Comment (makushline) [reply server/routes/spotify.js:169]: True, but it represents a single object in the database.

**Review by makushline — CHANGES_REQUESTED** [review]

**Review by makushline — APPROVED** [review]

**Pull request merged**